packages/block-library/src/legacy-widget/edit/preview.js

```diff
@@ -6,13 +6,44 @@ import classnames from 'classnames';
 /**
  * WordPress dependencies
  */
+import { useRefEffect } from '@wordpress/compose';
 import { addQueryArgs } from '@wordpress/url';
 import { useState } from '@wordpress/element';
 import { Placeholder, Spinner, Disabled } from '@wordpress/components';
 import { __ } from '@wordpress/i18n';
 
 export default function Preview( { idBase, instance, isVisible } ) {
-	const [ iframeHeight, setIframeHeight ] = useState( null );
+	const [ iframeHeight, setIframeHeight ] = useState();
+
+	// Resize the iframe on either the load event, or when the iframe becomes visible.
+	const ref = useRefEffect( ( iframe ) => {
+		function onChange() {
+			const boundingRect = iframe?.contentDocument?.body?.getBoundingClientRect();
+			if ( boundingRect ) {
+				// Include `top` in the height calculation to avoid the bottom
+				// of widget previews being cut-off. Most widgets have a
+				// heading at the top that has top margin, and the `height`
+				// alone doesn't take that margin into account.
+				setIframeHeight( boundingRect.top + boundingRect.height );
```

Comment:
I'm still seeing scrollbars when expanding a widget area though.

![image](https://user-images.githubusercontent.com/7753001/119922396-4b11d880-bfa2-11eb-8ded-104131a1a0ec.png)

Maybe we need to do something like this:

```js
const height = Math.max(
  boundingRect.top + boundingRect.height,
  iframe.contentDocument.documentElement.scrollHeight,
);
```

Reply:
Haha I was just popping in to say we could use `contentDocument.documentElement.scrollHeight` instead of the body boundingRect to simplify calculations and get rid of the scrollbar 😄

Reply:
scrollHeight is how it used to work, but it doesn't take into account margins.

Reply:
The idea was to measure the scrollHeight of the `html` element, so it includes the margins of all the elements inside, whereas with the `body` they sometimes collapse out of it.

Reply:
Ah yeah, sorry, I missed that bit of nuance.